File: smtk/bridge/exodus/testing/cxx/smtkReadFile.cxx
//=========================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//=========================================================================

#include "smtk/AutoInit.h"
#include "smtk/PublicPointerDefs.h"
#include "smtk/attribute/Attribute.h"
#include "smtk/attribute/FileItem.h"
#include "smtk/attribute/IntItem.h"
#include "smtk/attribute/System.h"
#include "smtk/model/EntityRef.h"
#include "smtk/model/EntityTypeBits.h"
#include "smtk/model/Manager.h"
#include "smtk/model/Model.h"
#include "smtk/model/Operator.h"

#include "smtk/common/testing/cxx/helpers.h"

#include <iostream>
#include <string>

// This macro ensures that exodus session is initialized properly on each platform
smtkComponentInitMacro(smtk_exodus_session)

  int main(int argc, char* argv[])
{
  // basic check info
  if (argc < 2)
  {
    std::cout << "Must provide input file as argument" << std::endl;
    return 1;
  }
  std::string modelPath(argv[1]);

  // Create model manager and exodus session
  smtk::model::ManagerPtr manager = smtk::model::Manager::create();
  auto session = manager->createSession("exodus");

  // Load model file
  smtk::model::OperatorPtr readOp = session.op("load smtk model");
  test(!!readOp, "No operator import-smtk-model.");

  readOp->specification()->findFile("filename")->setValue(modelPath);
  std::cout << "Importing " << modelPath << "\n";
  smtk::model::OperatorResult opresult = readOp->operate();
  test(opresult->findInt("outcome")->value() == smtk::model::OPERATION_SUCCEEDED,
    "Read operator failed");
  smtk::model::EntityRef entRef = opresult->findModelEntity("created")->value();
  smtk::model::Model model(entRef);

  // Check model validity
  test(model.isValid(), "Invalid model");

  // Check model geometry style
  test(model.geometryStyle() == smtk::model::DISCRETE, "Incorrect geometry style (not discrete)");

  test(model.session() == session, "Model should have parent session.");
  test(session.models<std::set<smtk::model::Model> >().find(model)->isValid(),
    "Session should own model.");

  session.close();

  //std::cout << "finis" << std::endl;
  return 0;
}

File: smtk/bridge/exodus/testing/python/smtkReadFile.py
#=============================================================================
#
#  Copyright (c) Kitware, Inc.
#  All rights reserved.
#  See LICENSE.txt for details.
#
#  This software is distributed WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#  PURPOSE.  See the above copyright notice for more information.
#
#=============================================================================
import os
import smtk
from smtk.simple import *
if smtk.wrappingProtocol() == 'pybind11':
    import smtk.model
    import smtk.bridge.exodus
import smtk.testing
import sys


class TestExodusSession(smtk.testing.TestCase):

    def testReadSMTK(self):
        import os
        smtk_file = os.path.join(
            smtk.testing.DATA_DIR, 'model', '3d', 'exodus', 'disk_out_ref.smtk')
        print 'smtk_file', smtk_file

        model_manager = smtk.model.Manager.create()
        sess = model_manager.createSession('exodus')
        sess.assignDefaultName()
        SetActiveSession(sess)

        LoadSMTKModel(smtk_file)
        result = GetLastResult()

        outcome = result.findInt('outcome').value(0)
        self.assertEqual(
            outcome,
            smtk.model.OPERATION_SUCCEEDED,
            'Import operation failed')

        model_ent = result.findModelEntity('created').value(0)
        model = smtk.model.Model(model_ent)
        self.assertTrue(model.isValid(), 'Imported model not valid')

        # Verify that the model is marked as discrete
        self.assertEqual(
            model.geometryStyle(), smtk.model.DISCRETE,
            'Expected discrete model, got {gs}'.format(gs=model.geometryStyle()))


if __name__ == '__main__':
    smtk.testing.process_arguments()
    smtk.testing.main()
